### ISSUE TITLE
doc request: describe how a lease transfer occurs after node failure

### ISSUE THREAD

**Issue by rafiss** [cockroachdb/docs#10913]
This issue is synthesizing a common request from the field. (See https://cockroachdb.zendesk.com/agent/tickets/7951 for a recent example.)

We currently have documentation about regular lease transfers here: https://www.cockroachlabs.com/docs/stable/architecture/replication-layer.html

But customers also are interested in the survivability of CockroachDB. Specifically, customers would like to know an **upper bound time** on how long a range will be unavailable after the leaseholder of the range suddenly dies.

https://github.com/cockroachdb/cockroach/pull/65195 is a recent PR that changes how a lease is transferred off of a dead node. It contains some bullet points about what is going on. This information should be included in our documentation.

```
    Wait 9s for liveness to expire
    Attempt to acquire the lease on a healthy node, this is rejected
    because lease acquisition can only happen on the raft leader
    The rejection unquiesces the range
    The nodes wait out the election timeout (3s) before trying to
    campaign
    A raft election occurs and a healthy node wins
    A lease transfer can now be processed by the newly elected leader
```

Also, the documentation should cover the special case of when the internal `liveness` range is colocated with some other lease for range "X". Does that make the lease transfer of "X" take even longer?

**Comment by rmloveland** [comment]
Thanks for reporting @rafiss - working on a docs PR now

re: this question

> Also, the documentation should cover the special case of when the internal liveness range is colocated with some other lease for range "X". Does that make the lease transfer of "X" take even longer?

@irfansharif do you mind helping with this question? looks like you were touching some things in `liveness.go` somewhat recently.  I read a bunch of things but can't claim to understand them.  At a high level the answer seems to be "yes, but this is pretty unlikely" since the node liveness range is apparently very early in the system keyspace "to limit the number of ranges which must use expiration-based leases" (per `constants.go`)

FWIW this seems very very edge-casey - the problem with putting super duper edge cases in docs (unless they are needed to solve a problem multiple users are hitting) is that every reader has to pay the cost of looking at that complexity, but perhaps only 1 reader actually wanted that info when having e.g. a support interaction with us, and the appropriate answer for the general user reading the docs may be "if you really really care about this level of detail, please see the code", even though for that 1 big user we go ahead and answer directly - but IDK!  It's possible I am misunderstanding something, which happens a lot! Curious what y'all think!

**Comment by rafiss** [comment]
> this is pretty unlikely since the node liveness range is apparently very early in the system keyspace "to limit the number of ranges which must use expiration-based leases"

A major motivation for this docs request was a report from a customer who observed a long period of inability to create any new connections to any node in the cluster when one other node was forcefully terminated and black-holed. It was actually easy to reproduce reliably. Note that the system ranges are also very early in the keyspace. During authentication user/password information needs to be looked from a system range. So if that system range is colocated with the liveness range, it made the period of inability-to-connect last longer.

> FWIW this seems very very edge-casey - the problem with putting super duper edge cases in docs (unless they are needed to solve a problem multiple users are hitting)

I don't feel like it's too edge-casey. Multiple customers have made docs requests for this area, according to our Customer Success team. cc @BramGruneir may be able to add more flavor about the importance of this documentation for customers.

**Comment by rmloveland** [comment]
> Note that the system ranges are also very early in the keyspace. During authentication user/password information needs to be looked from a system range. So if that system range is colocated with the liveness range, it made the period of inability-to-connect last longer.

Ah, ok - thanks for explaining this Rafi !  it makes sense.  in that case I will def write something up to address this specific issue

**Comment by rmloveland** [comment]
cc @ericharmeling @ianjevans since the fact we will have to explain "this is why your app has to wait to connect if any node goes down" is possibly further evidence re: us having code samples for connection retries https://github.com/cockroachdb/docs/issues/5470

also @rafiss OOC is there a CockroachDB issue for this?  would be good to have a mention from this issue.  Seems like "a long period of inability to create any new connections to any node in the cluster when one other node was forcefully terminated and black-holed ... easy to reproduce" is not good.  I tried [this search](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+auth+label%3AC-bug+-label%3AO-sentry+-label%3AO-roachtest+is%3Aopen+) and [this one](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+%22dead+node%22+label%3AC-bug+-label%3AO-sentry+-label%3AO-roachtest+is%3Aopen+) (and some others) before asking but I'm bad at finding things in the CRDB issues

**Comment by rafiss** [comment]
i replied on https://github.com/cockroachdb/docs/issues/10921

another motivation for docs on connection retries (https://github.com/cockroachdb/docs/issues/5470) relates to the CC upgrade process. customers have raised questions about errors they see when a node shuts down: https://cockroachlabs.atlassian.net/browse/CC-2756

**Comment by rmloveland** [comment]
(The "reference-y" portion of this was fixed, but see #10921 for the other part of this re: can't log in when node down)
